part2/stages/Functions/install-main

```diff
@@ -531,6 +536,95 @@ install_bootloader_from_dom0fs()
     return 0
 }
 
+create_efi_boot_entries()
+{
+    local DISK_DEV="/dev/${TARGET_DISK}"
+
+    do_cmd efibootmgr -w -L "OpenXT (safe graphics)" -l "\EFI\BOOT\BOOTX64.EFI" -c -d ${DISK_DEV} --part 1 -u "openxt-support-safe-graphics" >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't create EFI boot entry" >&2
+        return 1
+    fi
+
+    do_cmd efibootmgr -w -L "OpenXT (AMT)" -l "\EFI\BOOT\BOOTX64.EFI" -c -d ${DISK_DEV} --part 1 -u "openxt-support-amt" >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't create EFI boot entry" >&2
+        return 1
+    fi
+
+    do_cmd efibootmgr -w -L "OpenXT (console)" -l "\EFI\BOOT\BOOTX64.EFI" -c -d ${DISK_DEV} --part 1 -u "openxt-support-console" >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't create EFI boot entry" >&2
+        return 1
+    fi
+
+    do_cmd efibootmgr -w -L "OpenXT (console AMT)" -l "\EFI\BOOT\BOOTX64.EFI" -c -d ${DISK_DEV} --part 1 -u "openxt-support-console-amt" >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't create EFI boot entry" >&2
+        return 1
+    fi
+
+    do_cmd efibootmgr -w -L "OpenXT" -l "\EFI\BOOT\BOOTX64.EFI" -c -d ${DISK_DEV} --part 1 >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't create EFI boot entry" >&2
+        return 1
+    fi
+
+
+    # Remove duplicate entries (if any)
+    do_cmd efibootmgr -D >&2
+
+    return 0
+}
+
+copy_to_esp()
+{
+    local ESP=$(sfdisk -l -q | grep "EFI System" | awk '{ print $1 }')
+    if [ "${ESP}" == "" ]; then
+        echo "No EFI System Partition on disk ${DISK_DEV}" >&2
+        if [ -d /sys/firmware/efi/efivars ]; then
+            return 1
+        else
+            return 0
+        fi
+    fi
+
+    do_mount ${ESP} ${ESP_MOUNT} >&2
+    if [ "$?" != "0" ]; then
+        echo "Couldn't mount ${ESP}, already mounted?" >&2
+        return 1
+    fi
+
+    do_cmd mkdir -p ${ESP_MOUNT}/EFI/BOOT >&2
+    do_cmd cp -f ${DOM0_MOUNT}/boot/xen.efi \
+          ${DOM0_MOUNT}/boot/shimx64.efi \
+          ${DOM0_MOUNT}/boot/bzImage \
+          ${DOM0_MOUNT}/boot/initramfs.gz \
+          ${DOM0_MOUNT}/boot/microcode_intel.bin \
+          ${DOM0_MOUNT}/etc/xen/xenrefpolicy/policy/policy.24 \
+          ${DOM0_MOUNT}/usr/share/xenclient/bootloader/openxt.cfg \
+          ${ESP_MOUNT}/EFI/BOOT >&2
+
+    if [ "$?" != "0" ]; then
+        echo "Failed to copy files to ESP" >&2
+        do_cmd umount ${ESP_MOUNT}
+        return 1
+    fi
+    # Copy the shim to the EFI removable path to work around buggy firmwares
+    do_cmd cp -f ${ESP_MOUNT}/EFI/BOOT/shimx64.efi \
+          ${ESP_MOUNT}/EFI/BOOT/BOOTX64.EFI >&2
+    do_cmd sync >&2
+
+    do_cmd umount ${ESP_MOUNT} >&2
+
+    if [ -d /sys/firmware/efi/efivars ]; then
```

Comment:
Should we ever get here without an efivars directory?

Reply:
Yes, during an upgrade. The 32-bit dom0 can't talk to the UEFI firmware, so there is no efivarfs.